### PR TITLE
When source file is redirected, set the prototype correctly in node factory

### LINKED ISSUE #48039: Cannot read property 'get' of undefined – file.pragmas.get("jsximportsource") in getJSXImplicitImportBase
[This issue](https://github.com/microsoft/TypeScript/issues/43373)  was reported a while back, if I could re-open that issue, I would. The bug was accepted and assigned and put on a milestone, but eventually given up on because no one could provide a consistent reproduction:

https://github.com/microsoft/TypeScript/issues/43373

I'm opening this because **I have a consistent set of steps to reproduce the bug,** and I'm hoping someone can look into it.

In the original issue, the OP described the project being in a monorepo, and another commenter described having the same problem in a repo that relied upon another repo importing react components (I think). I mention that because the setup shown below is similar.  It's not a monorepo, but the issue occurs in a repo that relies upon importing components from elsewhere in the filesystem using `compilerOptions.paths` in the `tsconfig.json`, like this:

```json
{
  "compilerOptions": {
    "paths": {
      "@evans/*": ["../evans/src/components/*"]
    }
  }
}
```

And then in the repo components are imported like this:

```tsx
import CartComponent from '@evans/cart';
```

Here are the exact steps you can run to create the issue:

```bash
# set up supporting repo
git clone git@github.com:friends-library-dev/evans.git
cd evans
git fetch
git checkout graphql-refactor
git reset --hard f3879ecd
npm install

# set up main (erroring) repo
cd ../  # 👋  important! the repos must be side-by-side
git clone git@github.com:friends-library-dev/storybook-web.git
cd storybook-web
git reset --hard e8232dd
npm install

# produce the error
node --trace-deprecation --trace-warnings --unhandled-rejections=strict node_modules/.bin/build-storybook
```

Thanks so much for looking into this!

### PR DESCRIPTION
**Pull request by sheetalkamat**
Fixes #48039

### CODE REVIEW

**Comment by merceyz** [comment]
This should also close https://github.com/microsoft/TypeScript/issues/41717 since https://github.com/microsoft/TypeScript/pull/44554 was closed in favour of this one.

File: src/compiler/factory/nodeFactory.ts [hunk 1]
```diff
@@ -5275,7 +5275,7 @@ namespace ts {
             hasNoDefaultLib: boolean,
             libReferences: readonly FileReference[]
         ) {
-            const node = baseFactory.createBaseSourceFileNode(SyntaxKind.SourceFile) as Mutable<SourceFile>;
+            const node = (source.redirectInfo ? Object.create(source.redirectInfo.redirectTarget) : baseFactory.createBaseSourceFileNode(SyntaxKind.SourceFile)) as Mutable<SourceFile>;
```
Comment (DanielRosenwasser) [inline src/compiler/factory/nodeFactory.ts:5278]: Is setting the prototype pretty typical in factory functions and whatnot @rbuckton?

Comment (rbuckton) [reply src/compiler/factory/nodeFactory.ts:5278]: We generally don't, but apparently we use `Object.create` in program.ts for redirected source files.

**Review by rbuckton — APPROVED** [review]
I'm concerned with the fact redirection uses `Object.create` to wrap an existing source file because its easily overlooked. However, since this is already an existing behavior this change seems fine.

**Pull request merged**